sites/all/modules/ding_unilogin/ding_unilogin.module

```diff
@@ -0,0 +1,342 @@
+
+/**
+ * Implements hook_menu().
+ */
+function ding_unilogin_menu() {
+  $items = array();
+
+  // Administration config page.
+  $items['admin/config/ding/unilogin'] = array(
+    'title' => 'UNI•Login',
+    'description' => 'Configure UNI•Login access',
+    'page callback' => 'drupal_get_form',
+    'page arguments' => array('ding_unilogin_admin_settings_form'),
+    'access arguments' => array('configure unilogin'),
+    'file' => 'ding_unilogin.admin.inc',
+  );
+
+  $items['ding_unilogin/get/%'] = array(
+    'page callback' => 'ding_unilogin_url_json',
+    'page arguments' => array(2),
+    // Let anyone access.
+    'access callback' => TRUE,
+    'type' => MENU_CALLBACK,
+  );
+
+  return $items;
+}
+
+/**
+ * Implements hook_permission().
+ */
+function ding_unilogin_permission() {
+  return array(
+    'configure unilogin' => array(
+      'title' => t('Configure UNI•Login'),
+      'description' => t('Allow role to configure UNI•Login settings.'),
+    ),
+  );
+}
+
+/**
+ * Ajax callback for getting login URL.
+ *
+ * JS calls this with the current path of the browser window, in order to get
+ * a link to UNI•Login with the return path set properly. The form_alter that
+ * inserts the link cannot know what the path of the page is because it's
+ * loaded into the popup using AJAX.
+ */
+function ding_unilogin_url_json($path) {
+  drupal_json_output(array('url' => ding_unilogin_url(urldecode($path))));
+}
+
+/**
+ * Alter the login form.
+ *
+ * Add in link for logging in via UNI•Login.
+ *
+ * @param array $form
+ *   The form array to alter.
+ * @param array $form_state
+ *   The state of the form.
+ */
+function ding_unilogin_form_alter(&$form, &$form_state, $form_id) {
+  if (in_array($form_id, array(
+    'user_login',
+    'user_login_block',
+    'ding_user_authenticate_form',
+  ))) {
+    if (variable_get('ding_unilogin_id', '')) {
+      $form['unilogin'] = array(
+        '#attributes' => array(
+          'class' => array(
+            'unilogin-button',
+            // Hide the link, the JS will show it when it's set the proper URL.
+            'element-hidden',
+          ),
+        ),
+        '#type' => 'link',
+        '#title' => t('Log in via !service',
+                      array(
+                        '!service' => '<span class="unilogin-logo">UNI•Login</span>',
+                      )),
+        '#options' => array('html' => TRUE),
+        '#attached' => array(
+          'js' => array(
+            drupal_get_path('module', 'ding_unilogin') . '/js/ding_unilogin.js',
+          ),
+        ),
+      );
+    }
+  }
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter() for profile2_form().
+ *
+ * Hide the pincode change widget for UNI•Login users.
+ */
+function ding_unilogin_form_profile2_form_alter(&$form, &$form_state) {
+  if (ding_unilogin_authenticated()) {
+    if (isset($form['pass'])) {
+      $form['pass']['#access'] = FALSE;
+    }
+  }
+}
+
+/**
+ * Handle logging in returing UNI•Login users.
```

Comment:
If you have a link to some techinical documentation about how to properly validate a UNI-C login then I would appreciate a reference to it.